modules/hal_nordic/nrf_802154/sl_opensource/platform/nrf_802154_lp_timer_zephyr.c

```diff
@@ -99,99 +74,83 @@ void nrf_802154_lp_timer_deinit(void)
 	nrf_802154_clock_lfclk_stop();
 }
 
-void nrf_802154_lp_timer_critical_section_enter(void)
+uint64_t nrf_802154_platform_sl_lptimer_current_lpticks_get(void)
 {
-	nrf_802154_sl_mcu_critical_state_t state;
-
-	nrf_802154_sl_mcu_critical_enter(state);
-
-	if (!m_in_critical_section) {
-		(void)z_nrf_rtc_timer_compare_int_lock(m_rtc_channel);
-		m_in_critical_section = true;
-	}
-
-	nrf_802154_sl_mcu_critical_exit(state);
+	return z_nrf_rtc_timer_read();
 }
 
-void nrf_802154_lp_timer_critical_section_exit(void)
+uint64_t nrf_802154_platform_sl_lptimer_us_to_lpticks_convert(uint64_t us, bool round_up)
 {
-	nrf_802154_sl_mcu_critical_state_t state;
-
-	nrf_802154_sl_mcu_critical_enter(state);
-
-	m_in_critical_section = false;
-
-	z_nrf_rtc_timer_compare_int_unlock(m_rtc_channel, m_is_running);
-
-	nrf_802154_sl_mcu_critical_exit(state);
+	if (round_up) {
+		return NRF_802154_SL_US_TO_RTC_TICKS(us);
+	} else {
+		return NRF_802154_SL_US_TO_RTC_TICKS(us) - 1ULL;
+	}
 }
 
-uint32_t nrf_802154_lp_timer_time_get(void)
+uint64_t nrf_802154_platform_sl_lptimer_lpticks_to_us_convert(uint64_t lpticks)
 {
-	/* Please note that this function does not handle RTC overflow */
-	return NRF_802154_SL_RTC_TICKS_TO_US(z_nrf_rtc_timer_read());
+	/* Calculations are performed on uint64_t as it is safe to assume
+	 * overflow will not occur in any foreseeable future.
+	 */
+	return NRF_802154_SL_RTC_TICKS_TO_US(lpticks);
 }
 
-uint32_t nrf_802154_lp_timer_granularity_get(void)
+void nrf_802154_platform_sl_lptimer_schedule_at(uint64_t fire_lpticks)
 {
-	return NRF_802154_SL_US_PER_TICK;
-}
+	/* This function is not required to be reentrant, hence no critical section. */
+	if ((m_last_scheduled_time != fire_lpticks) || !m_is_running) {
+		m_last_scheduled_time = fire_lpticks;
+		m_is_running = true;
+		__DMB();
```

Comment:
What is this barrier and other barriers added in this PR trying to prevent? If `m_is_running` time does not restrict access to other variables (in other words is not a mutex or any other synchronization primitive), then the barrier is not needed.

Reply:
The above barrier might probably be removed. However, please note that the value of `m_is_running` through the [following line](https://github.com/nrfconnect/sdk-zephyr/pull/632/files#diff-589a6a1e1c76d9a63c44277064aa946de44e31702360d0a5ad34e696c86f1a33R102) affects whether `z_nrf_rtc_timer_set` is called. In my understanding, that indicates that some barriers introduced in this PR are needed (e.g. the one in `nrf_802154_platform_sl_lptimer_disable`).

Reply:
Looks like barriers are not needed. As for handling of `m_is_running` in critical sections, `__disable_irq` and `__enable_irq` have compiler barriers in them. The rest should be handled properly according to the C standard, eg. the one in `nrf_802154_platform_sl_lptimer_disable` is at the end of the function, hence an implicit `return`.

Reply:
We should have abandoned `m_is_running` earlier.